espnet/nets/pytorch_backend/conformer/encoder_layer.py

```diff
@@ -0,0 +1,155 @@
+        self.norm1 = LayerNorm(size)  # for the first FNN module
+        self.norm2 = LayerNorm(size)  # for MHA module
+        self.norm3 = LayerNorm(size)  # for CNN module
+        self.norm4 = LayerNorm(size)  # for the second FNN module
+        self.norm5 = LayerNorm(size)
+        self.conv_block = conv_block
+        self.dropout = nn.Dropout(dropout_rate)
+        self.size = size
+        self.normalize_before = normalize_before
+        self.concat_after = concat_after
+        self.rel_pos = rel_pos
+        self.macaron_style = macaron_style
+        if self.concat_after:
+            self.concat_linear = nn.Linear(size + size, size)
+
+    def forward(self, x_input, mask, cache=None):
+        """Compute encoded features.
+
+        :param torch.Tensor x_input: encoded source features, w/o pos_emb
+        (batch, max_time_in, size) or tuple((batch, max_time_in, size),
+        (1, max_time_in, size))
+        :param torch.Tensor mask: mask for x (batch, max_time_in)
+        :param torch.Tensor cache: cache for x (batch, max_time_in - 1, size)
+        :rtype: Tuple[torch.Tensor, torch.Tensor]
+        """
+        if self.rel_pos:
+            x, pos_emb = x_input[0], x_input[1]
+        else:
+            x = x_input
+
+        # whether to use macaron style
+        if self.macaron_style:
+            residual = x
+            if self.normalize_before:
+                x = self.norm1(x)
+            x = residual + 0.5 * self.dropout(self.feed_forward(x))
+            if not self.normalize_before:
+                x = self.norm1(x)
+
+        # multi-headed self-attention module
+        residual = x
+        if self.normalize_before:
+            x = self.norm2(x)
+
+        if cache is None:
+            x_q = x
+        else:
+            assert cache.shape == (x.shape[0], x.shape[1] - 1, self.size)
+            x_q = x[:, -1:, :]
+            residual = residual[:, -1:, :]
+            mask = None if mask is None else mask[:, -1:, :]
+
+        if self.rel_pos:
+            x_att = self.self_attn(x_q, x, x, pos_emb, mask)
+        else:
+            x_att = self.self_attn(x_q, x, x, mask)
+
+        if self.concat_after:
+            x_concat = torch.cat((x, x_att), dim=-1)
+            x = residual + self.concat_linear(x_concat)
+        else:
+            x = residual + self.dropout(x_att)
+        if not self.normalize_before:
+            x = self.norm2(x)
+
+        # convolution module
+        if self.conv_block is not None:
+            residual = x
+            if self.normalize_before:
+                x = self.norm3(x)
+            x = residual + self.dropout(self.conv_block(x))
+            if not self.normalize_before:
+                x = self.norm3(x)
+
+        # feed forward module
+        residual = x
+        if self.normalize_before:
+            x = self.norm4(x)
+        if self.macaron_style:
+            x = residual + 0.5 * self.dropout(self.feed_forward(x))
```

Comment:
IMPORTANT: this should be self.feed_forward2.

Reply:
Yes, it should be self.feed_forward2. Before initing the PR, I just shared the parameters of FNN modules and forgot to modify this part. I will change this part now.

Reply:
OK. Or It might be better to have such an option to share FFN parameters.

Reply:
Yes, if `marcon_style = False`, we only use `self.feed_forward2`.